File: dayFour/python/credit_card_validator.py
def convert_number_to_array(card_number):
    card_digits = []

    for digit in str(card_number):
        card_digits.append(int(digit))

    return card_digits


def get_card_type(card_digits):
    if len(card_digits) == 0:
        return "Unknown Card"

    if card_digits[0] == 4:
        return "Visa Card"

    if card_digits[0] == 5:
        return "MasterCard"

    if card_digits[0] == 6:
        return "Discover Card"

    if len(card_digits) > 1 and card_digits[0] == 3 and card_digits[1] == 7:
        return "American Express Card"

    return "Unknown Card"


def calculate_sum_of_doubled_even_position_digits(card_digits):
    doubled_digit_sum = 0

    for current_index in range(len(card_digits) - 2, -1, -2):
        doubled_digit = card_digits[current_index] * 2

        if doubled_digit > 9:
            tens_digit = doubled_digit // 10
            units_digit = doubled_digit % 10

            doubled_digit = tens_digit + units_digit

        doubled_digit_sum += doubled_digit

    return doubled_digit_sum


def calculate_sum_of_odd_position_digits(card_digits):
    odd_position_digit_sum = 0

    for current_index in range(len(card_digits) - 1, -1, -2):
        odd_position_digit_sum += card_digits[current_index]

    return odd_position_digit_sum


def is_valid_card(card_digits):
    doubled_digit_sum = calculate_sum_of_doubled_even_position_digits(card_digits)

    odd_position_digit_sum = calculate_sum_of_odd_position_digits(card_digits)

    total_sum = doubled_digit_sum + odd_position_digit_sum

    validity_status = total_sum % 10 == 0

    return validity_status

#def main():
#
#    credit_card_number = int(input("Enter credit card number: "))
#
#    card_digits = convert_number_to_array(credit_card_number)
#
#    card_type = get_card_type(card_digits)
#
#    validity_status = is_valid_card(card_digits)
#
#    print("Card Type:", card_type)
#
#    print("Validity Status:", validity_status)
#
#main()

File: dayFour/java/creditCardValidator/CreditCardValidator.java
import java.util.Scanner;

public class CreditCardValidator {

    public static int[] convertNumberToArray(long cardNumber) {

    long temporaryCardNumber = cardNumber;

    int count = 0;

    while (temporaryCardNumber > 0) {

        count++;

        temporaryCardNumber  = temporaryCardNumber / 10;

    }

    int[] cardDigits = new int[count];

    for (int digit = count - 1; digit >= 0; digit--) {

        cardDigits[digit] = (int)(cardNumber % 10);

        cardNumber = cardNumber / 10;

    }

    return cardDigits;
}

    public static String getCardType(int[] cardDigits) {

        if (cardDigits.length == 0) {

            return "Unknown Card";

        }

        if (cardDigits[0] == 4) {

            return "Visa Card";

        }

        if (cardDigits[0] == 5) {

            return "MasterCard";

        }

        if (cardDigits[0] == 6) {

            return "Discover Card";

        }

        if (cardDigits.length > 1

                && cardDigits[0] == 3

                && cardDigits[1] == 7) {

            return "American Express Card";

        }

        return "Unknown Card";

    }

    public static int calculateSumOfDoubledEvenPositionDigits(int[] cardDigits) {

        int doubledDigitSum = 0;

        for (int currentdigit = cardDigits.length - 2; currentdigit >= 0; currentdigit -= 2) {

            int doubledDigit = cardDigits[currentdigit] * 2;

            if (doubledDigit > 9) {

                int tensDigit = doubledDigit / 10;

                int unitsDigit = doubledDigit % 10;

                doubledDigit = tensDigit + unitsDigit;

            }

            doubledDigitSum += doubledDigit;
        }

        return doubledDigitSum;
    }

    public static int calculateSumOfOddPositionDigits(int[] cardDigits) {

        int oddPositionDigitSum = 0;

        for (int currentdigit = cardDigits.length - 1; currentdigit >= 0; currentdigit -= 2) {

            oddPositionDigitSum += cardDigits[currentdigit];

        }

        return oddPositionDigitSum;

    }


    public static boolean isValidCard(int[] cardDigits) {

        int doubledDigitSum = calculateSumOfDoubledEvenPositionDigits(cardDigits);

        int oddPositionDigitSum = calculateSumOfOddPositionDigits(cardDigits);

        int digitSum = doubledDigitSum + oddPositionDigitSum;

        return digitSum % 10 == 0;

    }
    
    public static void main(String[] args) {

        Scanner inputCollector = new Scanner(System.in);

        System.out.print("Enter credit card number: ");

        long creditCardNumber = inputCollector.nextLong();

        int[] cardDigits = convertNumberToArray(creditCardNumber);

        System.out.println("Card Type: " + getCardType(cardDigits));

        System.out.println("Validatity Status: " + isValidCard(cardDigits));

        

    }

}
